display_list/dl_op_spy.cc

```diff
@@ -0,0 +1,171 @@
+}
+void DlOpSpy::drawPoints(PointMode mode,
+                         uint32_t count,
+                         const SkPoint points[]) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && count > 0;
+}
+void DlOpSpy::drawVertices(const DlVertices* vertices, DlBlendMode mode) {
+  if (did_draw_ || mode == DlBlendMode::kDst) {
+    return;
+  }
+  did_draw_ =
+      dirty_ && mode != DlBlendMode::kClear && vertices->vertex_count() > 0;
+}
+// In theory, below drawImage methods can produce a transparent screen when a
+// transparent image is provided. The operation of determine whether an image is
+// transparent needs examine all the pixels in the image object, which is slow.
+// Drawing a completely transparent image is not a valid use case, thus, such
+// case is ignored.
+void DlOpSpy::drawImage(const sk_sp<DlImage> image,
+                        const SkPoint point,
+                        DlImageSampling sampling,
+                        bool render_with_attributes) {
+  did_draw_ = true;
+}
+void DlOpSpy::drawImageRect(const sk_sp<DlImage> image,
+                            const SkRect& src,
+                            const SkRect& dst,
+                            DlImageSampling sampling,
+                            bool render_with_attributes,
+                            SrcRectConstraint constraint) {
+  did_draw_ = true;
+}
+void DlOpSpy::drawImageNine(const sk_sp<DlImage> image,
+                            const SkIRect& center,
+                            const SkRect& dst,
+                            DlFilterMode filter,
+                            bool render_with_attributes) {
+  did_draw_ = true;
+}
+void DlOpSpy::drawAtlas(const sk_sp<DlImage> atlas,
+                        const SkRSXform xform[],
+                        const SkRect tex[],
+                        const DlColor colors[],
+                        int count,
+                        DlBlendMode mode,
+                        DlImageSampling sampling,
+                        const SkRect* cull_rect,
+                        bool render_with_attributes) {
+  did_draw_ = true;
+}
+void DlOpSpy::drawDisplayList(const sk_sp<DisplayList> display_list,
+                              SkScalar opacity) {
+  if (did_draw_ || opacity == 0) {
+    return;
+  }
+  DlOpSpy receiver;
+  display_list->Dispatch(receiver);
+  did_draw_ = receiver.DidDraw();
+}
+void DlOpSpy::drawTextBlob(const sk_sp<SkTextBlob> blob,
+                           SkScalar x,
+                           SkScalar y) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_;
+}
+void DlOpSpy::drawShadow(const SkPath& path,
+                         const DlColor color,
+                         const SkScalar elevation,
+                         bool transparent_occluder,
+                         SkScalar dpr) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = !color.isTransparent() && elevation > 0 && !path.isEmpty();
```

Comment:
I'm not sure about elevation == 0. I think it might also depend on transparent_occluder as well.

Reply:
what does `transparent_occluder` do?